The 'ticket_lifetime' option isn't documented.  RT#6680

Index: doc/admin.texinfo
===================================================================
--- doc/admin.texinfo	(revision 23799)
+++ doc/admin.texinfo	(working copy)
@@ -583,6 +583,11 @@
 fail if the client machine does not have a keytab.  The default for the
 flag is @value{DefaultVerifyApReqNofail}.
 
+@itemx ticket_lifetime
+The value of this tag is the default lifetime for
+initial tickets.  The default value for the tag is
+@value{DefaultTktLifetime}.
+
 @itemx renew_lifetime
 The value of this tag is the default renewable lifetime for
 initial tickets.  The default value for the tag is
Index: src/config-files/krb5.conf.M
===================================================================
--- src/config-files/krb5.conf.M	(revision 23799)
+++ src/config-files/krb5.conf.M	(working copy)
@@ -220,6 +220,10 @@
 fail if the client machine does not have a keytab.  The default for the
 flag is false.
 
+.IP ticket_lifetime
+The value of this tag is the default lifetime for initial tickets.  The
+default value for the tag is 1 day (1d).
+
 .IP renew_lifetime
 The value of this tag is the default renewable lifetime for initial
 tickets.  The default value for the tag is 0.
